File: reverse-integer/string-to-integer-atoi/myAtoi.c
int myAtoi(char * s){
    int i;
    int sign;
    int num;

    i = 0;
    sign = 1;
    num = 0;
    while (*s <= 32 && *s >= 13)
        s++;
    if (*s == '-')
    {
        s++;
        sign *= -1;
    }
    else if (*s == '+')
    {
        s++;
    }
    while (*s >= '0' && *s <= '9'){
        int digit = *s++ - '0';
        if (num > INT_MAX / 10 || (num == INT_MAX / 10 && digit > 7)) {
            return (sign == 1) ? INT_MAX : INT_MIN;
        }
        num = ((10 * num) + digit);
    }
    return (num * sign);
}

File: reverse-integer/string-to-integer-atoi/myAtoi.go
import "math"

func myAtoi(s string) int {
    num := 0
    sign := 1
    i := 0

    for i < len(s) && s[i] == ' ' {
        i++
    }

    if i < len(s) && (s[i] == '-' || s[i] == '+') {
        if s[i] == '-' {
            sign = -1
        }
        i++
    }

    for i < len(s) && s[i] >= '0' && s[i] <= '9' {
        digit := int(s[i] - '0')
        if num > math.MaxInt32/10 || (num == math.MaxInt32/10 && digit > 7) {
            if sign == 1 {
                return math.MaxInt32
            } else {
                return math.MinInt32
            }
        }

        num = num*10 + digit
        i++
    }
    return num * sign
}
